File: src/api/_fixtures/offers.js
const COMP = [
  { id: '1', name: 'PrintMagic Pro', initials: 'PM' },
  { id: '2', name: 'DTFworld',       initials: 'DW' },
  { id: '3', name: 'PrintZone',      initials: 'PZ' },
  { id: '4', name: 'ThreadBeast',    initials: 'TB' },
  { id: '5', name: 'VividPrints',    initials: 'VP' },
]

export const OFFER_HEX = {
  Discount:       '#EF4444',
  Bundle:         '#8B5CF6',
  'Free Shipping':'#3B82F6',
  BOGO:           '#F97316',
  'Limited Time': '#F59E0B',
  Guarantee:      '#22C55E',
  Other:          '#94A3B8',
}

const TEXTS = {
  Discount: [
    '40% Off Custom Prints — Today Only',
    'Save $20 on Orders Over $100',
    '25% Off Your First Order',
    'Bulk Discount: Up to 60% Off',
    'Flash Sale: $0.39/sheet for 24 Hours',
  ],
  Bundle: [
    'Buy Any 3, Get 1 Free',
    'Gang Sheet Bundle — Save 30%',
    'Starter Pack: 50 Sheets + Free Setup',
    'Pro Bundle: Design + Print Combo',
  ],
  'Free Shipping': [
    'Free Shipping on All Orders Over $50',
    'Free Express Delivery This Week',
    'Free Shipping — No Minimum Required',
  ],
  BOGO: [
    'Buy One Gang Sheet, Get One 50% Off',
    'BOGO: 2-for-1 on All Custom Transfers',
    'Double Your Order — Pay for One',
  ],
  'Limited Time': [
    '48-Hour Flash Sale — Ends Tonight',
    'Weekend Only: Prices Slashed',
    'Last Chance: Sale Ends Sunday',
  ],
  Guarantee: [
    '100% Satisfaction or Full Money Back',
    'Reprint Guarantee — No Questions Asked',
  ],
}

const DESCS = {
  Discount:       'A direct price reduction designed to lower the perceived barrier to purchase and incentivise immediate action.',
  Bundle:         'Groups multiple products or quantities together at a reduced combined price, increasing average order value.',
  'Free Shipping':'Removes the additional cost of delivery, one of the top reasons customers abandon carts at checkout.',
  BOGO:           'Buy-one-get-one mechanics reward repeat purchase intent while creating a strong perception of value.',
  'Limited Time': 'Scarcity and urgency mechanics that push prospects over the decision threshold before an offer window closes.',
  Guarantee:      'Risk-reversal promise that reduces purchase anxiety and builds trust by standing behind product quality.',
}

const HOOK_POOL = ['Pain', 'Urgency', 'Benefit', 'Curiosity', 'Social Proof', 'Trust', 'How To']
const COMP_CYCLE = [0, 2, 1, 3, 0, 4, 1, 2, 3, 4, 0, 1, 2, 3, 0, 4, 1, 2, 3, 0, 1, 4]

// 22 rows — top 6 types then sub-variants
const SEQUENCE = [
  ['Discount', 0],       ['Bundle', 0],         ['Free Shipping', 0],  ['BOGO', 0],
  ['Limited Time', 0],   ['Guarantee', 0],       ['Discount', 1],       ['Bundle', 1],
  ['Free Shipping', 1],  ['BOGO', 1],            ['Discount', 2],       ['Limited Time', 1],
  ['Bundle', 2],         ['Discount', 3],        ['BOGO', 2],           ['Free Shipping', 2],
  ['Guarantee', 1],      ['Discount', 4],        ['Bundle', 3],         ['Limited Time', 2],
  ['Discount', 0],       ['BOGO', 0],
]

const MENTIONS_LIST   = [412, 287, 245, 178, 156, 84, 97, 76, 68, 61, 54, 48, 43, 39, 35, 32, 29, 26, 22, 18, 15, 12]
const CONFIDENCE_LIST = [74, 82, 77, 79, 71, 88, 69, 85, 73, 76, 67, 70, 80, 65, 74, 72, 91, 63, 78, 68, 66, 75]
const TRENDING_LIST   = [8.4, -2.1, 5.7, 28.6, 12.3, -1.4, 3.9, -5.2, 17.1, 6.8, -3.7, 9.2, -1.8, 4.4, 22.5, -8.1, 2.6, 11.3, -4.9, 7.7, 1.2, -6.3]

function buildRow(type, ti, i) {
  const text    = TEXTS[type][ti % TEXTS[type].length]
  const hex     = OFFER_HEX[type] ?? '#94A3B8'
  const cStart  = COMP_CYCLE[i] ?? (i % COMP.length)
  const compAll = [...COMP.slice(cStart), ...COMP.slice(0, cStart)]
  const compN   = 1 + (i % 3)
  const extra   = i % 2
  return {
    id:               String(i + 1),
    rank:             i + 1,
    text,
    description:      DESCS[type],
    type,
    mentions:         MENTIONS_LIST[i],
    avg_confidence:   CONFIDENCE_LIST[i],
    trending:         TRENDING_LIST[i],
    competitors:      compAll.slice(0, compN),
    extra_competitors: extra,
    example_ads:      Array.from({ length: 4 }, (_, j) =>
      `https://placehold.co/80x80/${hex.replace('#', '')}/ffffff?text=${type[0]}${j + 1}`
    ),
    first_seen:       new Date(2026, 1, 1 + i * 9).toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' }),
    related_hooks:    [HOOK_POOL[i % HOOK_POOL.length], HOOK_POOL[(i + 2) % HOOK_POOL.length]],
  }
}

export const offersTable = SEQUENCE.map(([type, ti], i) => buildRow(type, ti, i))

export const offersSummary = {
  total_unique:           89,
  total_mentions:         1624,
  mentions_trend:         9.7,
  top_offer_type:         'Discount',
  top_offer_type_pct:     38.4,
  discount_heavy_share:   38.4,
  trending_offer:         { text: 'BOGO: 2-for-1 on All Custom Transfers', type: 'BOGO', pct: 28.6 },
}

export const offersTypeDist = [
  { name: 'Discount',       value: 624, pct: 38.4 },
  { name: 'Bundle',         value: 287, pct: 17.7 },
  { name: 'Free Shipping',  value: 245, pct: 15.1 },
  { name: 'BOGO',           value: 178, pct: 11.0 },
  { name: 'Limited Time',   value: 156, pct:  9.6 },
  { name: 'Guarantee',      value:  84, pct:  5.2 },
  { name: 'Other',          value:  50, pct:  3.1 },
]

export const offersPerformance = [
  { type: 'Guarantee',     avg_score: 84, color: '#22C55E' },
  { type: 'Bundle',        avg_score: 79, color: '#8B5CF6' },
  { type: 'BOGO',          avg_score: 76, color: '#F97316' },
  { type: 'Limited Time',  avg_score: 73, color: '#F59E0B' },
  { type: 'Free Shipping', avg_score: 70, color: '#3B82F6' },
  { type: 'Discount',      avg_score: 67, color: '#EF4444' },
]

const TREND_DATES = ['May 2', 'May 3', 'May 4', 'May 5', 'May 6', 'May 7', 'May 8']
export const offersTrend = TREND_DATES.map((date, i) => ({
  date,
  Discount:       Math.round(56 + Math.sin(i * 0.7)       * 8 + i * 0.5),
  Bundle:         Math.round(38 + Math.sin(i * 0.5 + 1)   * 6 + i * 0.4),
  'Free Shipping':Math.round(32 + Math.sin(i * 1.1 + 2)   * 5 + i * 0.3),
  BOGO:           Math.round(22 + Math.sin(i * 0.9 + 0.5) * 4 + i * 0.6),
}))

// legacy alias
export const offersList = offersTable

File: backend/app/routers/offers.py
"""
Offer Library endpoints — aggregations over ad_analyses.offer_type field.
"""

from datetime import datetime, timezone, timedelta
from typing import List

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func, distinct
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.models.ad import Ad
from app.models.ad_analysis import AdAnalysis
from app.models.competitor import Competitor
from app.schemas.offer import (
    OffersSummary,
    TrendingOffer,
    OfferTypeDistItem,
    OfferPerformanceItem,
    OfferTrendPoint,
    OfferRow,
    OfferCompetitorRef,
)


router = APIRouter(prefix="/offers", tags=["offers"])


OFFER_COLORS = {
    "Discount": "#EF4444",
    "Bundle": "#8B5CF6",
    "Free Shipping": "#3B82F6",
    "BOGO": "#F97316",
    "Limited Time": "#F59E0B",
    "Free Trial": "#22C55E",
    "Guarantee": "#22C55E",
    "Other": "#94A3B8",
    "None": "#94A3B8",
}

OFFER_DESCRIPTIONS = {
    "Discount": "A direct price reduction designed to lower the perceived barrier to purchase and incentivise immediate action.",
    "Bundle": "Groups multiple products or quantities together at a reduced combined price, increasing average order value.",
    "Free Shipping": "Removes the additional cost of delivery, one of the top reasons customers abandon carts at checkout.",
    "BOGO": "Buy-one-get-one mechanics reward repeat purchase intent while creating a strong perception of value.",
    "Limited Time": "Scarcity and urgency mechanics that push prospects over the decision threshold before an offer window closes.",
    "Free Trial": "Reduces commitment risk by letting prospects experience the product before paying.",
    "Guarantee": "Risk-reversal promise that reduces purchase anxiety and builds trust by standing behind product quality.",
}


def _initials(name: str) -> str:
    if not name:
        return "?"
    words = name.split()
    if len(words) >= 2:
        return (words[0][0] + words[1][0]).upper()
    return name[:2].upper()


# ---------- Endpoints ----------

@router.get("/summary", response_model=OffersSummary)
async def get_offers_summary(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Aggregate KPIs for the Offers Library top cards."""
    # Exclude None/null offer types
    unique_stmt = select(func.count(distinct(AdAnalysis.offer_type))).where(
        AdAnalysis.offer_type.is_not(None),
        AdAnalysis.offer_type != "None",
    )
    total_unique = (await db.execute(unique_stmt)).scalar() or 0

    total_stmt = select(func.count(AdAnalysis.id)).where(
        AdAnalysis.offer_type.is_not(None),
        AdAnalysis.offer_type != "None",
    )
    total_mentions = (await db.execute(total_stmt)).scalar() or 0

    # Top offer type
    top_stmt = (
        select(AdAnalysis.offer_type, func.count(AdAnalysis.id).label("c"))
        .where(
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
        )
        .group_by(AdAnalysis.offer_type)
        .order_by(func.count(AdAnalysis.id).desc())
        .limit(1)
    )
    row = (await db.execute(top_stmt)).first()
    top_offer_type = row[0] if row else None
    top_count = row[1] if row else 0
    top_pct = round(top_count / total_mentions * 100, 1) if total_mentions else 0.0

    # Discount-heavy share = % of mentions that are "Discount"
    disc_stmt = select(func.count(AdAnalysis.id)).where(
        AdAnalysis.offer_type == "Discount"
    )
    disc_count = (await db.execute(disc_stmt)).scalar() or 0
    disc_share = round(disc_count / total_mentions * 100, 1) if total_mentions else 0.0

    # Trending offer — most-used offer text in last 7 days
    week_ago = datetime.now(timezone.utc) - timedelta(days=7)
    trend_stmt = (
        select(
            AdAnalysis.offer_value,
            AdAnalysis.offer_type,
            func.count(AdAnalysis.id).label("c"),
        )
        .where(
            AdAnalysis.offer_value.is_not(None),
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
            AdAnalysis.analyzed_at >= week_ago,
        )
        .group_by(AdAnalysis.offer_value, AdAnalysis.offer_type)
        .order_by(func.count(AdAnalysis.id).desc())
        .limit(1)
    )
    trend_row = (await db.execute(trend_stmt)).first()
    trending = (
        TrendingOffer(
            text=trend_row[0],
            type=trend_row[1],
            pct=round(trend_row[2] / max(total_mentions, 1) * 100, 1),
        )
        if trend_row else None
    )

    return OffersSummary(
        total_unique=total_unique,
        total_mentions=total_mentions,
        mentions_trend=0.0,
        top_offer_type=top_offer_type,
        top_offer_type_pct=top_pct,
        discount_heavy_share=disc_share,
        trending_offer=trending,
    )


@router.get("/type-dist", response_model=List[OfferTypeDistItem])
async def get_offers_type_distribution(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Offer type distribution for donut chart."""
    stmt = (
        select(AdAnalysis.offer_type, func.count(AdAnalysis.id).label("c"))
        .where(
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
        )
        .group_by(AdAnalysis.offer_type)
        .order_by(func.count(AdAnalysis.id).desc())
    )
    rows = (await db.execute(stmt)).all()
    total = sum(r[1] for r in rows)

    return [
        OfferTypeDistItem(
            name=r[0],
            value=r[1],
            pct=round(r[1] / total * 100, 1) if total else 0.0,
        )
        for r in rows
    ]


@router.get("/performance", response_model=List[OfferPerformanceItem])
async def get_offers_performance(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Avg confidence score per offer type — bar chart."""
    stmt = (
        select(
            AdAnalysis.offer_type,
            func.avg(AdAnalysis.confidence_score).label("avg_score"),
        )
        .where(
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
        )
        .group_by(AdAnalysis.offer_type)
        .order_by(func.avg(AdAnalysis.confidence_score).desc())
    )
    rows = (await db.execute(stmt)).all()

    return [
        OfferPerformanceItem(
            type=r[0],
            avg_score=round(float(r[1]), 1),
            color=OFFER_COLORS.get(r[0], "#94A3B8"),
        )
        for r in rows
    ]


@router.get("/trend", response_model=List[OfferTrendPoint])
async def get_offers_trend(
    days: int = Query(7, ge=1, le=90),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Offer mentions over time."""
    start_date = datetime.now(timezone.utc) - timedelta(days=days - 1)
    start_date = start_date.replace(hour=0, minute=0, second=0, microsecond=0)

    stmt = (
        select(
            func.date_trunc("day", AdAnalysis.analyzed_at).label("day"),
            AdAnalysis.offer_type,
            func.count(AdAnalysis.id).label("c"),
        )
        .where(
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
            AdAnalysis.analyzed_at >= start_date,
        )
        .group_by("day", AdAnalysis.offer_type)
        .order_by("day")
    )
    rows = (await db.execute(stmt)).all()

    points: dict = {}
    for i in range(days):
        d = (start_date + timedelta(days=i)).date()
        date_str = d.strftime("%b %d")
        points[date_str] = {
            "date": date_str,
            "Discount": 0,
            "Bundle": 0,
            "Free Shipping": 0,
            "BOGO": 0,
        }

    for row in rows:
        day_date = row[0].date() if hasattr(row[0], "date") else row[0]
        date_str = day_date.strftime("%b %d")
        offer = row[1]
        count = row[2]
        if date_str in points and offer in ("Discount", "Bundle", "Free Shipping", "BOGO"):
            points[date_str][offer] = count

    # Convert dict-based points to model instances using aliased field
    result = []
    for p in points.values():
        result.append(OfferTrendPoint(
            date=p["date"],
            Discount=p["Discount"],
            Bundle=p["Bundle"],
            **{"Free Shipping": p["Free Shipping"]},
            BOGO=p["BOGO"],
        ))
    return result


@router.get("", response_model=List[OfferRow])
async def list_offers(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List all detected offers with stats, sorted by mentions desc."""
    stmt = (
        select(
            AdAnalysis.offer_value,
            AdAnalysis.offer_type,
            func.count(AdAnalysis.id).label("mentions"),
            func.avg(AdAnalysis.confidence_score).label("avg_conf"),
            func.min(AdAnalysis.analyzed_at).label("first_seen"),
        )
        .where(
            AdAnalysis.offer_type.is_not(None),
            AdAnalysis.offer_type != "None",
        )
        .group_by(AdAnalysis.offer_value, AdAnalysis.offer_type)
        .order_by(func.count(AdAnalysis.id).desc())
    )
    rows = (await db.execute(stmt)).all()

    result: List[OfferRow] = []
    for rank, r in enumerate(rows, start=1):
        offer_value, offer_type, mentions, avg_conf, first_seen = r

        # Competitors using this offer
        comp_stmt = (
            select(Competitor.id, Competitor.name)
            .join(Ad, Ad.competitor_id == Competitor.id)
            .join(AdAnalysis, AdAnalysis.ad_id == Ad.id)
            .where(AdAnalysis.offer_type == offer_type)
            .distinct()
        )
        comp_rows = (await db.execute(comp_stmt)).all()
        all_comps = [
            OfferCompetitorRef(id=str(c[0]), name=c[1], initials=_initials(c[1]))
            for c in comp_rows
        ]
        shown = all_comps[:3]
        extra = max(0, len(all_comps) - 3)

        # Example ad media URLs
        media_stmt = (
            select(Ad.media_url)
            .join(AdAnalysis, AdAnalysis.ad_id == Ad.id)
            .where(AdAnalysis.offer_type == offer_type)
            .limit(3)
        )
        media_rows = (await db.execute(media_stmt)).all()
        example_ads = [m[0] for m in media_rows if m[0]]

        first_seen_str = first_seen.strftime("%b %d, %Y") if first_seen else None

        # Display text: prefer offer_value if present, else the type label
        text = offer_value if offer_value else offer_type

        result.append(
            OfferRow(
                id=str(rank),
                rank=rank,
                text=text,
                description=OFFER_DESCRIPTIONS.get(offer_type, ""),
                type=offer_type,
                mentions=mentions,
                avg_confidence=round(float(avg_conf), 1),
                trending=0.0,
                competitors=shown,
                extra_competitors=extra,
                example_ads=example_ads,
                first_seen=first_seen_str,
                related_hooks=[],
            )
        )

    return result